### ISSUE TITLE
scatter_ supporting different reduction modes

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#22378]
## 🚀 Feature

Currently, we have `scatter_` and `scatter_add_`.

However, in use-cases such as Graph Neural Networks, it's common to wanting to do other reductions than add.
As a case in point, @rusty1s has written https://github.com/rusty1s/pytorch_scatter which has `add`, `sub`, `mean`, `max`, `min`, `std`, `mul`, `div`.

The kernels are there, and they are minor variations of the existing scatter kernel, [see here](https://github.com/rusty1s/pytorch_scatter/blob/master/cuda/scatter_kernel.cu).

We should provide this in PyTorch natively, ideally has a `reduce='add'` keyword argument in `scatter_`, or if one wants it to be more explicit, `scatter_reduce_(..., mode='mean')`

## Motivation

Graph Neural Nets are becoming pretty common.

**Comment by rusty1s** [comment]
Looking forward to seeing this in native PyTorch. I'm not sure if this is possible at all, but I get a lot of requests regarding deterministic scatter executions. There is also [this thread](https://discuss.pytorch.org/t/possible-solution-to-the-reproducibility-issues-using-scatter-add-operation/48989) in the official PyTorch forum which makes suggestions on how this could be achieved. What are your thoughts on that?

**Comment by fmassa** [comment]
I believe in numpy the (almost) equivalent would be [`np.ufunc.at`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ufunc.at.html), where ufunc can be [any of the universal functions]( https://docs.scipy.org/doc/numpy/reference/ufuncs.html)

Might be good having a look to see if we would want to use the numpy API in this case

**Comment by vadimkantorov** [comment]
Custom reduction for all operations with `out` argument specified would be super-cool. But probably many of them don’t have `atomicAdd` equivalents in CUDA, hence big rewrites of kernels...

**Comment by t-vi** [comment]
Last I looked `atomicAdd` wasn't that bad to implement manually, but it's not deterministic. Deterministic implementation takes some effort (as discussed in the thread linked above). Nonetheless a bit of templating it might work out well with a bit of templating.

**Comment by v0dro** [comment]
I'm coming up with a proposal to implement this and will share it soon.

**Comment by nikitaved** [comment]
@v0dro, there is a scatter_gather kernel which uses TensorIterator and hence supports multithreading in full but atomic operations. Can be trivially used for some, primarily pointwise operations. Very relevant issues are [31662](https://github.com/pytorch/pytorch/pull/31662), and especially [31776](https://github.com/pytorch/pytorch/issues/31776). Also, expect some issues with backward methods, you will need to implement some of them from scratch, which is also a non-trivial task, given that some of the methods, like `scatter/scatter_add` have incorrect backward methods, see [31776](https://github.com/pytorch/pytorch/issues/31776) for more details.

**Comment by v0dro** [comment]
Yeah this is a big issue! Especially given the implementation of determinism... but I think can be done piece by piece.

@nikitaved thank you very much for the information :)

**Comment by nikitaved** [comment]
Yeah, once you have atomic operations implemented, the existing TensorIterator can be trivially used. Then forward is easy to implement. The issues will arise in backward especially in cases where sources contribute to single locations in results. You have to keep the set of indices contributing to the result in forward, and use them to zero out some of them in backward. There is also a dilemma between determinism/nondeterminism there...

CC @ngimel

**Comment by v0dro** [comment]
Since this is a pretty big issue I decided to cut it down into more simple and independent sub-parts, at least some of which can be done in parallel:

1. Implement atomic operations for various binary operations and parallelize the (now serial) `scatter_add` that was implemented in #31662. (#33065 )
2. Non-deterministic scatter reduction algorithms for various scatter operations (target API would be something like `dest.scatter_(dim, index, src, reduce="sum")`.
3. Offer a choice between deterministic and non-deterministic algorithms for the reduction depending on whether the user is okay with such behaviour (target API would be something like `dest.scatter_(dim, index, src, reduce="sum", deterministic=True)`.
4. Update documentation to reflect the new behaviour with determinism and non-unique indices.

@ngimel @nikitaved do have a look at the sub-parts which don't have associated issues and if there is any need for more granularity in the planning. I can go ahead and open issues for each sub-part later.

**Comment by accelsao** [comment]
Does `non-deterministic algorithms for the reduction` means the output is *correct* by may not be the same each time?

**Comment by gchanan** [comment]
I don't think you have to add both deterministic and non-deterministic algorithms at once, if you want to reduce the scope of the problem.

Also, I don't think I would (publicly) expose the deterministic ones (assuming they are slower).  See https://github.com/pytorch/pytorch/issues/15359, where the proposals are around a more global flag, which seems to make more sense for the problem space (I generally don't care whether one specific algorithm is deterministic).  So I'd probably just make the deterministic variants prefixed with "_" to start, e.g. "_scatter_mean_deterministic".

**Comment by v0dro** [comment]
> Does non-deterministic algorithms for the reduction means the output is correct by may not be the same each time?

Yes that's right.

> I don't think you have to add both deterministic and non-deterministic algorithms at once, if you want to reduce the scope of the problem.

I'm not. I'm breaking down the issue into sub-issues that can be tackled separately. Deterministic implementations are a separate issue.

> See #15359, where the proposals are around a more global flag, which seems to make more sense for the problem space (I generally don't care whether one specific algorithm is deterministic).
My reasoning for adding a `deterministic` option is that providing a global option would involve a survey of methods that might be deterministic other than the scatter reduction methods. Therefore the changes required for such an option will need to have a wider scope than we have now.

**Comment by gchanan** [comment]
The issue suggests having functionality like scatter_div, but I don't see that in pytorch_scatter any longer.  @rusty1s was that functionality removed? why?

**Comment by gchanan** [comment]
or maybe those just aren't documented?

**Comment by rusty1s** [comment]
You are right. They got removed by the major rewrite in >= 2.0. I have plans to add `scatter_mul` back in, but `scatter_div` will likely not be implemented since it does not really have a use-case. If you desperately need this functionality, you need to install `torch-scatter<2`.

**Comment by v0dro** [comment]
After much deliberation and revisions, scatter reductions for both CPU and CUDA are now complete. https://github.com/pytorch/pytorch/pull/41977 for CUDA, https://github.com/pytorch/pytorch/pull/36447 for CPU. Both implementations exist for addition and multiplication since @ngimel determined the subtraction and division to be unnecessary. The CPU implementation is always deterministic since the threading methodology ensures that the operations always happen in the same order. However the CUDA implementation in non-deterministic due to lack of control over the order in which the operations take place.

That makes a deterministic implementation of scatter reduction for CUDA the only part of this issue that remains to be implemented. However, in issue https://github.com/pytorch/pytorch/issues/23151 a separate API for deterministic scatter operations was proposed, which still remains an open issue. Whether such an API is still necessary is a matter debate, since a global setting for deterministic/non-deterministic behaviour for pytorch algorithms has been proposed in the past.

Given that all functionality for scatter reduction except determinism for CUDA has been implemented, I propose closing this issue and moving further discussion to https://github.com/pytorch/pytorch/issues/23151 or opening a brand new issue for this purpose.

CC: @rgommers

**Comment by ezyang** [comment]
sgtm

**Comment by ezyang** [comment]
@ngimel is going to respond here

**Comment by ngimel** [comment]
To have parity with pytorch_scatter, we still need scatter_min, scatter_max and scatter_mean. Also, it would be good to have benchmarks comparing performance in pytorch core and pytorch_scatter. @rusty1s is there a good set of benchmarks for scatter operations?

**Comment by rusty1s** [comment]
I have written some benchmarks for `torch-scatter`, see [here](https://github.com/rusty1s/pytorch_scatter/blob/master/benchmark/scatter_segment.py). It can download some sparse matrices from the [SuiteSparse Matrix Collection](https://sparse.tamu.edu/) and performs scatter operations for different feature dimensionalites. I hope that this is useful to you :)

**Comment by v0dro** [comment]
Thank you for your feedback. So here's my thoughts on the future course of action:

1. Write a benchmarking suite in the benchmarks/ folder comparing various scatter modes from pytorch and pytorch-scatter. Reuse tests from @rusty1s 's comment above. Make this a separate issue.
2. Expand current scatter reduction with support for min, max and mean for both CPU and CUDA. Incorporate this into the benchmarks above. Make this is a separate issue.

Since the core issues that this issue had to address have been addressed, I think we can close this issue and move the discussion to the dedicated issues. I think it would lead to less bloat and confine this issue to building the scatter reduction infrastructure, which has now been built and can be easily extended to support other scatter reduction modes.

@ngimel thoughts?

**Comment by rgommers** [comment]
> Write a benchmarking suite in the `benchmarks/` folder comparing various scatter modes from pytorch and pytorch-scatter. Reuse tests from @rusty1s 's comment above. Make this a separate issue.

I think writing benchmarks and committing them is the right idea, but those should only benchmark the PyTorch implementation. The comparison with `torch-scatter` seems better to do as a one-off. That's the way comparisons with NumPy, TensorFlow and other libraries are mostly handled too - the README in `benchmarks/` claims comparisons to other frameworks can be added, but in practice it's rarely done, and especially for small libraries like `torch-scatter` that may change quickly, I'm not sure that's useful to add.

> Expand current scatter reduction with support for min, max and mean for both CPU and CUDA. Incorporate this into the benchmarks above. Make this is a separate issue.

Adding those modes needs doing indeed. I'd prefer no more separate issues. Let's just focus on doing this all under this umbrella issue. Too many separate issues were created already, and it's confusing more than helpful. Adding the new modes should be straightforward now that the other PRs are merged.

**Comment by v0dro** [comment]
@ngimel question about behaviour of scatter reduction when using `min` or `max` reduction options:

In the pytorch-scatter, there is no need to supply the `input` tensor since it is initilzed to `0` internally and therefore something like `max` only concerns the `src` tensor. Therefore, for pytorch-scatter the code turns out as:

``` python
src = torch.tensor([[6, 8], [1,2]], device=device, dtype=dtype)  
index = torch.tensor([[1], [2]], device=device, dtype=torch.long) 
scatter(src, index, dim=0, reduce='max')                                                                                                                                                                                      
# Out[25]: 
# tensor([[0., 0.],
#        [0., 0.],
#        [6., 8.]], dtype=torch.float64)
```

However, in pytorch, scatter expects another `input` tensor that might or might not be zero-initilized. In this case, for add we simply `+=` the result of the scatter addition to the respective element in the `input` tensor and `*=` for multiply.

When doing this for `max` or `min`, should the value in the `input` tensor be replaced with the result of the scatter reduction of should it be `max/min(input_value, scatter_reduced_value)`? Also, for `mean`, should the respective `input` value be included in the result when computing the mean?

**Comment by rgommers** [comment]
Good question @v0dro, I think there's a problem here - including in the initial feature request. It conceptually doesn't make sense to have
```
scatter(..., reduce='operation')
```
where `'operation'` can be both an element-wise operation or something that already is a reduction. `'add'`, `'subtract'`, `'multiply'` are implemented now, those are element-wise. `'mean'`, `'std'` are reduction operations. For `'min'`, `'max'`, there's two options - element-wise comparison or reduction (numpy has both, see `np.min` and `np.minimum`).

In `torch-scatter` the `scatter` function only has reductions: https://github.com/rusty1s/pytorch_scatter/blob/master/torch_scatter/scatter.py#L130.

One option for an API, which is consistent with `torch-scatter`, would have been:
- `scatter(..., reduce='any_reduction_operation')`   # can include `'mean'`, `'min'`, `'max'`, `'sum'`, `'std'`
- separate methods for inplace operators: `scatter_add` (exists), `scatter_multiply`, `scatter_subtract`

An alternative, which is consistent with what the `reduce=` keyword does now, is:
- `scatter(..., reduce='any_elementwise_operation')`:  now `'add'`, `'multiple'`, `'subtract'`, but there's many others possible
- separate methods for reductions: `scatter_mean`, `scatter_max`, `scatter_min`  

This is actually consistent with numpy, where ufuncs have a `reduce` method (e.g. `np.add.reduce` is equivalent to `np.sum`). One downside of course is that `scatter_add` already exists, which is inconsistent with the current choice in `reduce=`. 

Looking at @soumith's feature request, which also suggests `scatter_reduce_`, perhaps the most consistent alternative would be:
- `scatter_(..., mode='elementwise-op')`
- `scatter_reduce_((..., mode='reduction-op')`
This would require deprecating the just introduced `reduce=`, but it's be a nice and concise API.

Finally a note on the current `reduce='add'`: as the docs example clearly shows, nothing is being "reduced" there, the `reduce='add'` simply means using `+=` rather than `=` when assigning the result of the actual scatter operation (which applies `index` to `src`).

@ngimel, @rusty1s - thoughts?

**Comment by mruberry** [comment]
@rgommers and @ngimel and I had a chance to discuss this offline recently. Here's what we're thinking:

- torch.Tensor.scatter_ has a "reduce" argument that, ideally, would take the name of a reduction and perform the reduction on the set of {value in self/dst, values "scattered" to that location from src}. The current names can stay for BC-compatibility, but we should probably add the names of the reductions they correspond to. That is, "add" corresponds to "sum," and "multiply" corresponds to "prod." "sub" doesn't have a corresponding reduction, but the tensor can be negated before the sum operation is applied to achieve the same effect.
- "pytorch_scatter," the library developed by @rusty1s, has a function called "scatter" with different, but related, functionality. This functionality can be implemented separately as a different function with a name like "scattered." 

Separately, it would be nice to support a more complete universal function (AKA "ufunc") interface.

**Comment by stas00** [comment]
Any chance pytorch could add support for "sum", "mean", "max" or "min", in addition to `mul` and `add` implementations?

It'd help HF transformers not to need to depend on `pytorch_scatter` which isn't trivial to build and requires us to bug  `pytorch_scatter` to often build binary wheels.

Thank you!

**Comment by ezyang** [comment]
I think this is the right spot. Thanks for reminding us about this again!

**Comment by vadimkantorov** [comment]
Another practical usecase: computing bounding boxes of image segments (i.e. aggregate min/max coordinates per segment label)

**Comment by stas00** [comment]
ping?

**Comment by cpuhrsch** [comment]
@stas00 - I'm writing up an RFC for this as we speak. The issue is figuring out the right API since pytorch_scatters' scatter is different from pytorch's scatter. As a basic example for this the [documentation of PyTorch's scatter](https://pytorch.org/docs/master/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_) explicitly says that repeat indices results in inconsistent behavior, whereas pytorch_scatter's scatter seems to explicitly allow them just based [on the example alone](https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html). Some approaches to resolve API issues were already pointed out by commenters above so I'll include them.

**Comment by cpuhrsch** [comment]
How about this:

We take [torch.scatter](https://pytorch.org/docs/master/generated/torch.scatter.html?highlight=scatter#torch.scatter), make src optional (src = None) and add a reduce flag (sum, prod, mean, std).

If src is None, the input is the source, if src is given we preserve the current behavior. We deprecate reduce='add' and reduce='multiply' for scatter_ and replace them by 'sum' and 'prod'. Then we're unblocked to add 'mean', 'std', 'min', 'max' for torch.scatter, because only the chosen values from the source need to be of that reduction and we return newly allocated values. Otherwise the user will need to fill the input with identity values to get this behavior, which is very annoying for operations like std or or max. If someone wants the output to be a particular given size we can also add an out flag to torch.scatter. The new signature then would be
```torch.scatter(input, dim, index, src=None, reduce=None, out=None)```.

This shouldn't cause any BC issues and the new functionality added to torch.scatter should be able to encompass those of [torch_scatter.scatter](https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html).

**Comment by cpuhrsch** [comment]
@ngimel @bhosmer @mruberry and I talked offline and decided to introduce a scatter_reduce (similar to @rgommers proposal), but with the semantics of torch_scatter's [scatter](https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html).

**Comment by tvercaut** [comment]
I would also be interested in this, especially the `scatter_mean`. Of course it is possible to use `scatter_add` twice for this, once of the tensor we want mean-reduced, once on a `ones_like` of it, and divide the two but that is not super elegant and certainly suboptimal.

**Comment by cpuhrsch** [comment]
@tvercaut, we have [`scatter_reduce`](https://pytorch.org/docs/master/generated/torch.Tensor.scatter_reduce_.html#torch.Tensor.scatter_reduce_) which also has a `mean` option.

**Comment by tvercaut** [comment]
Thanks for pointing out and sorry for missing it. I had looked at [`scatter`](https://pytorch.org/docs/1.13/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_) which according to the doc provides a `reduce` option (currently) limited to `add` and `multiply`. It might be good for the `scatter` documentation to point to further reduction options being offered in `scatter_reduce`.

**Comment by cpuhrsch** [comment]
@tvercaut - That's a good point. We'll update the documentation and also add a link. Sorry for the churn. Please let us know if this works for you.

**Comment by function2-llx** [comment]
Hello, it seems that reduce for `mean` divide count+1, is this right?

To reproduce:

```python
# Python 3.10.9, PyTorch 1.13.1.post200
print(torch.zeros(3).scatter_reduce_(dim=0, index=torch.tensor([0, 1, 1, 2, 2, 2]), src=torch.ones(6), reduce='mean'))
# tensor([0.5000, 0.6667, 0.7500])
```

**Comment by rusty1s** [comment]
You can control this behavior via the `include_self` option.
